addons/hr_holidays/models/resource.py

```diff
@@ -1,10 +1,136 @@
+        gto_end = date_utils.end_of(gto['date_to'], 'day')
+        leave_tz = timezone(leave['employee_id']['resource_id']['tz'])
+
+        if gto_start <= leave_start\
+                and gto_end > leave_start\
+                and gto_end < leave_end:
+            leave.write({
+                'date_from': leave_tz.localize(gto_end + timedelta(seconds=1))\
+                        .astimezone(UTC).replace(tzinfo=None)
+            })
+            return self.env['hr.leave'].sudo()
+        if gto_start > leave_start\
+                and gto_end < leave_end:
+            copys = {
+                'date_from': leave['date_from'],
+                'date_to': leave_tz.localize(gto_start - timedelta(seconds=1))\
+                        .astimezone(UTC).replace(tzinfo=None)
+            }
+            leave.write({
+                'date_from': leave_tz.localize(gto_end + timedelta(seconds=1))\
+                        .astimezone(UTC).replace(tzinfo=None)
+            })
+            return leave.copy(copys)
+        if gto_start > leave_start\
+                and gto_start < leave_end\
+                and gto_end >= leave_end:
+            leave.write({
+                'date_to': leave_tz.localize(gto_start - timedelta(seconds=1))\
+                        .astimezone(UTC).replace(tzinfo=None)
+            })
+            return self.env['hr.leave'].sudo()
+
+    def _split_leave(self, leave, time_domain_dict):
+        new_leaves = self.env['hr.leave'].sudo()
+        for record in sorted(
+                filter(lambda r: r['date_to'] > leave['date_from'] and r['date_from'] < leave['date_to'], time_domain_dict),
+                key=lambda r: r['date_from']):
+            new_leave = self._split_leave_on_gto(leave, record)
+            if new_leave:
+                new_leaves |= new_leave
+        return new_leaves
+
+    def _reevaluate_leaves(self, time_domain_dict):
+        if time_domain_dict:
+            domain = self._get_domain(time_domain_dict)
+            leaves = self.env['hr.leave'].search(domain)
+            if not leaves:
+                return
+
+            previous_durations = [leave.number_of_days for leave in leaves]
+            previous_states = [leave.state for leave in leaves]
+            leaves.sudo().write({
+                'state': 'draft',
+            })
+            leaves._compute_number_of_days()
+            for previous_duration, leave, state in zip(previous_durations, leaves, previous_states):
+                duration_difference = previous_duration - leave.number_of_days
+                if duration_difference > 0 and leave['holiday_allocation_id']:
+                    leave.message_post(body=_("Due to a change in global time offs, you have been granted %s day(s) back", duration_difference), subtype_id=1)
+                if leave.number_of_days > previous_duration\
+                        and leave['holiday_status_id']['name'] not in ['Sick Time Off']:
```

Comment:
We need to find a better solution than that, this will not work with any other configuration than the default one, and I am pretty sure it's a lot of them.